monitor/cert_submitter.go

```diff
@@ -0,0 +1,143 @@
+}
+
+// run starts a goroutine that calls submitCertificate, sleeps for
+// certSubmitInterval and then repeats forever.
+func (c *certSubmitter) run() {
+	go func() {
+		for {
+			c.submitCertificate()
+			c.logger.Printf("Sleeping for %s before next certificate submission\n",
+				c.certSubmitInterval)
+			c.clk.Sleep(c.certSubmitInterval)
+		}
+	}()
+}
+
+// submitCertificate issues a certificate with the certSubmitter's
+// certIssuer/certIssuerKey and submits it to a monitored log's add-chain
+// endpoint. The latency of the submission is tracked in the
+// `cert_submit_latency` prometheus histogram. If the submission fails, or the
+// returned SCT is invalid the `cert_submit_results` prometheus countervec is
+// incremented with a "fail" status tag. If the submission succeeds the
+// `cert_submit_results` prometheus countervec is incremented with a "ok" status
+// tag. An SCT is considered invalid if the signature does not validate, or if
+// the timestamp is too far in the future or the past (controlled by
+// `requiredSCTFreshness`).
+func (c *certSubmitter) submitCertificate() {
+	c.logger.Printf("Submitting certificate to %q\n", c.logURI)
+
+	cert, err := pki.IssueTestCertificate(c.certIssuerKey, c.certIssuer, c.clk)
+	if err != nil {
+		// This should not occur and if it does we should abort hard
+		panic(fmt.Sprintf("!!! Error issuing certificate: %s\n", err.Error()))
+	}
+
+	// Because ct-woodpecker issues directly off of a fake root the "chain" only
+	// contains the leaf certificate we minted in issueCertificate()
+	chain := []ct.ASN1Cert{
+		ct.ASN1Cert{Data: cert.Raw},
+	}
+
+	latencyLabels := prometheus.Labels{"uri": c.logURI}
+	failLabels := prometheus.Labels{"uri": c.logURI, "status": "fail"}
+	successLabels := prometheus.Labels{"uri": c.logURI, "status": "ok"}
+
+	start := c.clk.Now()
+	ctx, cancel := context.WithTimeout(context.Background(), submitTimeout)
+	defer cancel()
+	sct, err := c.client.AddChain(ctx, chain)
+	elapsed := c.clk.Since(start)
+	c.stats.certSubmitLatency.With(latencyLabels).Observe(elapsed.Seconds())
+
+	if err != nil {
+		c.logger.Printf("!!! Error submitting certificate to %q: %s\n", c.logURI, err.Error())
+		c.stats.certSubmitResults.With(failLabels).Inc()
+		return
+	}
+
+	ts := time.Unix(0, int64(sct.Timestamp)*int64(time.Millisecond))
+	sctAge := c.clk.Since(ts)
+
+	// Check that the SCT's timestamp is within an allowable tolerance into the
+	// future and the past. The SCT's signature & log ID have already been verified by
+	// `m.client.AddChain()`
+	if sctAge > requiredSCTFreshness {
+		c.logger.Printf("!!! Error submitting certificate to %q: returned SCT timestamp signed %s in the future (expected < %s)",
+			c.logURI, sctAge, requiredSCTFreshness)
+		c.stats.certSubmitResults.With(failLabels).Inc()
+		return
+	} else if sctAge < -requiredSCTFreshness {
+		c.logger.Printf("!!! Error submitting certificate to %q: returned SCT timestamp signed %s in the past (expected > %s)",
```

Comment:
This "expected" clause should have a `<`. The overall thing might be clearer if you changed the if statement to say `-sctAge < requiredSCTFreshness`. Then both the minus sign and the angle bracket would match what is being logged.